src/interface/swarm.cpp

```diff
@@ -472,6 +479,111 @@ void Swarm::Defrag() {
   max_active_index_ = num_active_ - 1;
 }
 
+///
+/// Routine to sort particles by cell. Updates internal swarm variables:
+///  cellSorted_: 1D Per-cell sorted array of swarm memory indices
+///  (SwarmKey::swarm_index_) cellSortedBegin_: Per-cell array of starting indices in
+///  cellSorted_ cellSortedNumber_: Per-cell array of number of particles in each cell
+///
+void Swarm::SortParticlesByCell() {
+  auto pmb = GetBlockPointer();
+
+  auto &x = Get<Real>("x").Get();
+  auto &y = Get<Real>("y").Get();
+  auto &z = Get<Real>("z").Get();
+
+  const int nx1 = pmb->cellbounds.ncellsi(IndexDomain::entire);
+  const int nx2 = pmb->cellbounds.ncellsj(IndexDomain::entire);
+  const int nx3 = pmb->cellbounds.ncellsk(IndexDomain::entire);
+
+  auto cellSorted = cellSorted_;
+  int ncells = pmb->cellbounds.GetTotal(IndexDomain::entire);
+  int num_active = num_active_;
+  int max_active_index = max_active_index_;
+
+  // Allocate data if necessary
+  if (cellSortedBegin_.GetDim(1) == 0) {
+    cellSortedBegin_ = ParArrayND<int>("cellSortedBegin_", nx3, nx2, nx1);
+    cellSortedNumber_ = ParArrayND<int>("cellSortedNumber_", nx3, nx2, nx1);
+  }
+  auto cellSortedBegin = cellSortedBegin_;
+  auto cellSortedNumber = cellSortedNumber_;
+  auto swarm_d = GetDeviceContext();
+
+  // Write an unsorted list
+  pmb->par_for(
+      "Write unsorted list", 0, max_active_index_, KOKKOS_LAMBDA(const int n) {
+        int i, j, k;
+        swarm_d.Xtoijk(x(n), y(n), z(n), i, j, k);
+        const int cell_idx_1d = i + nx1 * (j + nx2 * k);
+        cellSorted(n) = SwarmKey(cell_idx_1d, n);
+      });
+
+  sort(cellSorted, SwarmKeyComparator(), 0, max_active_index);
```

Comment:
This actually came out really nice, didn't it?

Reply:
The wages of letting another library launch kernels for you :)